lib/device.c

```diff
@@ -387,6 +387,34 @@ int sync_save_tracks(const struct sync_device *d)
 
 #ifndef SYNC_PLAYER
 
+static int send_set_key_cmd(const struct sync_device *d, const struct sync_track *t, int row, float val, enum key_type type)
+{
+	unsigned char cmd = SET_KEY, _type = type;
+	uint32_t _track, _row;
+	union {
+		float f;
+		uint32_t i;
+	} v;
+
+	v.f = val;
+	v.i = htonl(v.i);
+
+	_track = find_track(d, t->name);
+	if (_track < 0)
+		return -1;
```

Comment:
It's a bit backwards to perform a binary-search for a struct that we already have. The problem is of course that you need the index rather than the pointer, and that's not trivially queriablefrom here. Perhaps we should make it trivially queriable, by storing the index in `struct sync_track` instead of having to search over and over again?

The error-condition should trigger at all here, so I would assert it instead; the only case I could imagine this happening is in some user-after-free or passing an invalid pointer... Neither are worth handling gracefully, IMO...

Reply:
Tell me about it, it's not even a binary search, it's an O(N) linear lookup in find_track().

However, it's not like demo-side set_value() is going to happen that much, and I was just keeping the changeset small using the existing structures and functions as-is.

We can always optimize that stuff later, no?

Reply:
Yeah, we can optimize it later. I still think the point about the error-handling stands, though?

Reply:
@kusma hey sorry for the delay, been busy with non-demo-stuff and forgot all about this.  Force-pushed with the assert instead of error return.